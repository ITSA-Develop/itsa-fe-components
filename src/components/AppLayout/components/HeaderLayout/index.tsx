import { ActiveNotificationIcon, NotificationIcon, PinIcon, UserIcon } from '@/assets/icons';
import { DropdownCustomLabel } from '@/components/DropdownCustomLabel';
import { DropdownIcon } from '@/components/DropdownIcon';
import { Image } from '@/components/Image';
import { LOGO_DIMENSIONS, ROUTES_IMAGES } from '@/constants';
import { useSidebarStore } from '@/hooks';
import { ISelectOptionDropdownButton } from '@/interfaces';
import { useAppLayoutStore } from '@/store/appLayout.store';
import { MenuUnfoldOutlined, SettingOutlined } from '@ant-design/icons';
import { Button, Dropdown, MenuProps } from 'antd';

export interface HeaderLayoutProps {
	loadingHeader: boolean;
	notifications?: MenuProps;
	userActions?: MenuProps;
	logo?: string;
}

export const HeaderLayout = ({
	loadingHeader,
	notifications = { items: [] },
	userActions = { items: [] },
	logo = '',
}: HeaderLayoutProps) => {
	const { setCurrentModule, setCurrentAgency } = useAppLayoutStore();
	const { collapsed, setCollapsed } = useSidebarStore();
	const { agencies } = useAppLayoutStore();
	const { currentAgency } = useAppLayoutStore();
	const { currentModule } = useAppLayoutStore();
	const { modulesAgency } = useAppLayoutStore();
	const { userName } = useAppLayoutStore();
	const { userRole } = useAppLayoutStore();

	const isActiveNotifications = Boolean(notifications?.items?.length);
	const isActiveUserActions = Boolean(userActions?.items?.length);

	const modulesData: ISelectOptionDropdownButton = {
		items: modulesAgency.map(m => ({
			key: m.id.toString(),
			label: m.name,
			onClick: (id: string) => handleSetCurrentModule(id),
		})),
	};

	const agenciesData: ISelectOptionDropdownButton = {
		items: agencies.map(a => ({
			key: a.id.toString(),
			label: a.name,
			onClick: (id: string) => handleSetCurrentAgency(id),
		})),
	};

	const handleSetCurrentModule = (moduleId: string) => {
		const module = modulesAgency.find(m => m.id.toString() === moduleId);
		if (module) {
			setCurrentModule(module);
		}
	};

	const handleSetCurrentAgency = (agencyId: string) => {
		const agency = agencies.find(a => a.id.toString() === agencyId);
		if (agency) {
			setCurrentAgency(agency);
		}
	};

	return (
		<header className="h-16">
			<div className="flex flex-row text-white-100 items-center justify-between w-full rounded-xl h-16 pr-4 pl-6 bg-primary-700">
				<div className="w-full flex flex-row items-center justify-start gap-4">
					{collapsed && (
						<Button
							type="text"
							icon={<MenuUnfoldOutlined className="text-white-100" />}
							onClick={() => setCollapsed(!collapsed)}
						/>
					)}
					<Image
						imgPath={logo || ROUTES_IMAGES.companyLogo}
						width={LOGO_DIMENSIONS.HEADER_WIDTH}
						height={LOGO_DIMENSIONS.HEADER_HEIGHT}
						alt="Logo"
						errorLabel="Importadora Tomebamba"
					/>
				</div>
				<div className="hidden sm:flex flex-row w-auto items-center gap-4">
					<div className="flex md:hidden flex-row items-center gap-4">
						<DropdownIcon
							options={modulesData}
							loading={loadingHeader}
							icon={<SettingOutlined className="text-white-100 w-4 h-4" />}
							onChange={handleSetCurrentModule}
						/>
						<DropdownIcon
							options={agenciesData}
							loading={loadingHeader}
							icon={<PinIcon className="text-white-100 w-4 h-4" />}
							onChange={handleSetCurrentAgency}
						/>
					</div>

					<div className="hidden md:flex flex-row items-center gap-4">
						<DropdownCustomLabel
							defaultValue={currentModule?.id?.toString()}
							options={modulesData}
							loading={loadingHeader}
							icon={<SettingOutlined className="text-white-100 w-4 h-4" />}
							emptyLabel="Sin mÃ³dulos asignados"
							onChange={handleSetCurrentModule}
						/>
						<DropdownCustomLabel
							defaultValue={currentAgency?.id?.toString()}
							options={agenciesData}
							loading={loadingHeader}
							icon={<PinIcon className="text-white-100 w-4 h-4" />}
							emptyLabel="Sin agencias asignadas"
							onChange={handleSetCurrentAgency}
						/>
						<div className="flex flex-col">
							<span className="text-4 whitespace-nowrap">{userName ?? ''}</span>
							<span className="text-primary-900 font-bold text-end text-xs whitespace-nowrap">{userRole ?? ''}</span>
						</div>
					</div>
				</div>
				<div className="flex flex-row pl-4">
					<Dropdown menu={userActions || { items: [] }} placement="bottomRight" disabled={!isActiveUserActions}>
						<Button type="text" icon={<UserIcon className="text-white-100 w-6 h-6" />} />
					</Dropdown>
					<Dropdown menu={notifications || { items: [] }} placement="bottomRight" disabled={!isActiveNotifications}>
						<Button
							type="text"
							icon={
								isActiveNotifications ? (
									<ActiveNotificationIcon className="text-white-100 w-6 h-6" />
								) : (
									<NotificationIcon className="text-white-100 w-6 h-6" />
								)
							}
						/>
					</Dropdown>
				</div>
			</div>
		</header>
	);
};
