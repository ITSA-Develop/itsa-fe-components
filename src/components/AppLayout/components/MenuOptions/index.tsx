import { TExtendedMenuItem, TMenuMode } from '@/types';
import { Menu, MenuProps } from 'antd';
import { SyncOutlined } from '@ant-design/icons';
import React from 'react';

export interface MenuOptionsProps {
	collapsed: boolean;
	currentPath: string;
	onClickOptionMenu: (info: { key: string; item: TExtendedMenuItem }) => void;
	items?: TExtendedMenuItem[];
	mode?: TMenuMode;
	openKeys?: string[];
	loading?: boolean;
	onOpenKeysChange?: (openKeys: string[]) => void;
}

const MenuOptions: React.FC<MenuOptionsProps> = ({
	collapsed,
	currentPath = '',
	onClickOptionMenu,
	items = [],
	mode = 'inline',
	openKeys = [],
	onOpenKeysChange,
	loading = false,
}) => {
	const handleMenuClick: MenuProps['onClick'] = info => {
		const findMenuItem = (menuItems: TExtendedMenuItem[], key: string): TExtendedMenuItem | null => {
			for (const item of menuItems) {
				if (item.key === key) {
					return item;
				}
				if ('children' in item && item.children) {
					const found = findMenuItem(item.children as TExtendedMenuItem[], key);
					if (found) return found;
				}
			}
			return null;
		};

		const clickedItem = findMenuItem(items, info.key);
		if (clickedItem && onClickOptionMenu) {
			onClickOptionMenu({ key: info.key, item: clickedItem });
		}
	};

	return (
		<div className="menu-options">
			{loading ? (
				<div className="flex flex-col items-center justify-center h-full pt-10 gap-2">
					<SyncOutlined spin className='text-gray-400' />
					<small className='text-gray-400'>Cargando men√∫</small>
				</div>
			) : (
				<Menu
					mode={mode}
					inlineCollapsed={collapsed}
					items={items}
					className="overflow-auto scrollbar-none border-none min-w-full"
					defaultSelectedKeys={[currentPath]}
					openKeys={openKeys}
					onOpenChange={onOpenKeysChange}
					onClick={handleMenuClick}
				/>
			)}
		</div>
	);
};

export default MenuOptions;
