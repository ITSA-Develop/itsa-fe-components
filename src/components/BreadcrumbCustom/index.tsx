import { Button } from 'antd';
import { LeftOutlined } from '@ant-design/icons';

export interface IBreadcrumbCustomProps {
	title: string;
	description: string;
	action: () => void;
}
export const BreadcrumbCustom = ({ title, description, action }: IBreadcrumbCustomProps) => {
	return (
		<div className="flex flex-row items-center gap-2">
			<Button onClick={action} type="text" icon={<LeftOutlined className="font-bold" style={{ fontSize: '20px' }} />} />
			<div className="flex flex-row items-center gap-2">
				<h4 className="font-bold text-lg text-black-100">{title}</h4>
				<span className="text-black-100">|</span>
				<small className="text-sm text-black-100">{description}</small>
			</div>
		</div>
	);
};
