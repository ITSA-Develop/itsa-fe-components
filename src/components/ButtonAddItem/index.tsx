import { Button, ButtonProps } from 'antd';
import { PlusOutlined } from '@ant-design/icons';

interface IButtonAddItemProps extends ButtonProps {
	label: string;
	onClick: () => void;
}

export const ButtonAddItem = ({ label, onClick,  ...props }: IButtonAddItemProps) => {
	return (
		<Button
            className={`h-6 bg-white !border !border-gray-900 text-gray-900 hover:!border-gray-900 hover:!bg-gray-50 hover:!text-gray-900 rounded-lg`}
			onClick={onClick}
			icon={<PlusOutlined className='text-gray-900 text-xs' />}
			{...props}
		>
			{label}
		</Button>
	);
};
