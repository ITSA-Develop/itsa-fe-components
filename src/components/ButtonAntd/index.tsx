import { Button as AntButton, ButtonProps } from 'antd';
import { ReactNode } from 'react';

export interface IButtonProps extends ButtonProps {
	children: ReactNode;
}

export const ButtonAntd = ({ children, ...rest }: IButtonProps) => {
	return <AntButton {...rest}>{children}</AntButton>;
};
