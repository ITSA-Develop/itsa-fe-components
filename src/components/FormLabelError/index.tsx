export interface IFormLabelErrorProps {
	label: string;
	htmlFor?: string;
	id?: string;
}

export const FormLabelError = ({ label, htmlFor, id }: IFormLabelErrorProps) => {
	return (
		<label htmlFor={htmlFor} id={id} className="text-red-500">
			{label}
		</label>
	);
};
