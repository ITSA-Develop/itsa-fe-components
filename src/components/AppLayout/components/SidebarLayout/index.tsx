import MenuOptions from '@/components/AppLayout/components/MenuOptions';
import { getAllMenuKeys, transformModuleToMenuData } from '@/helpers';
import { filterMenuItems } from '@/helpers/menu/menuDataTransformer';
import { useSidebarStore } from '@/hooks';
import { useAppLayoutStore } from '@/store/appLayout.store';
import { TExtendedMenuItem, TMenuMode } from '@/types';
import { DoubleLeftOutlined, SearchOutlined } from '@ant-design/icons';
import { Button, Input, Layout } from 'antd';
import { Content } from 'antd/es/layout/layout';
import { ReactNode, useEffect, useMemo } from 'react';
export interface SidebarLayoutProps {
	children: ReactNode;
	onClickOptionMenu: (info: { key: string; item: TExtendedMenuItem }) => void;
	width?: number;
	currentPath?: string;
	modeSidebar?: TMenuMode;
}
export const SidebarLayout = ({
	children,
	onClickOptionMenu,
	width = 235,
	currentPath = '',
	modeSidebar = 'inline',
}: SidebarLayoutProps) => {
	const currentModule  = useAppLayoutStore((state) => state.currentModule);

	const { collapsed, setCollapsed, searchTerm, setSearchTerm, openKeys, setOpenKeys } = useSidebarStore();
	const menuData = useMemo(() => {
		if (!currentModule) {
			return [];
		}
		const originalMenuData = transformModuleToMenuData(currentModule);

		if (!searchTerm.trim()) {
			return originalMenuData;
		}

		return filterMenuItems(originalMenuData, searchTerm);
	}, [currentModule, searchTerm]);

	useEffect(() => {
		if (searchTerm.length > 0) {
			const allKeys = getAllMenuKeys(menuData);
			setOpenKeys(allKeys);
		} else {
			setOpenKeys([]);
		}
	}, [searchTerm, menuData, setOpenKeys]);

	return (
		<Layout hasSider className="gap-2">
			{!collapsed && (
				<div className="flex flex-col pt-3 rounded-lg bg-gray-200" style={{ width: width }}>
					<div className="flex items-center justify-center pr-3 pl-3">
						<Input
							placeholder="Buscar en el menÃº"
							className="rounded-lg text-sm"
							suffix={<SearchOutlined className="text-gray-300" />}
							defaultValue={searchTerm}
							onChange={e => setSearchTerm(e.target.value)}
						/>
					</div>
					<div className="flex-1 overflow-y-auto scrollbar-none h-full max-w-full">
						<MenuOptions
							items={menuData}
							collapsed={collapsed}
							onClickOptionMenu={onClickOptionMenu}
							currentPath={currentPath || ''}
							mode={modeSidebar}
							openKeys={openKeys}
							onOpenKeysChange={setOpenKeys}
						/>
					</div>
					<div className="w-full flex justify-end pr-3 pl-3">
						<Button
							type="link"
							onClick={() => setCollapsed(true)}
							icon={<DoubleLeftOutlined className="text-gray-400" />}
						/>
					</div>
				</div>
			)}
			<Layout className="rounded-lg">
				<Content className="bg-white-100 overflow-auto rounded-lg pt-3 pb-2 pr-4 pl-4">{children}</Content>
			</Layout>
		</Layout>
	);
};
