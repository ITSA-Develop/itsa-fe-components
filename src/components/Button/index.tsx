import { ButtonAntd } from '../ButtonAntd';
import { ReactNode } from 'react';

export interface IButtonProps {
	size?: 'small' | 'middle' | 'large';
	type?: 'primary' | 'secondary';
	title?: ReactNode;
	disabled?: boolean;
	onClick?: () => void;
	/** Cuando es true en tipo secondary, aplica estilo gris (#595959) */
	default?: boolean;
}

export const Button = (props: IButtonProps) => {
	const { size = 'small', type = 'primary', title, disabled = false, onClick } = props;
	const sizeClass = size === 'small' ? 'itsa-btn--sm' : size === 'middle' ? 'itsa-btn--md' : 'itsa-btn--lg';
	const variantClass = type === 'primary' ? 'itsa-btn--primary' : 'itsa-btn--secondary';
	const disabledClass = disabled ? 'itsa-btn--disabled' : '';
	const defaultSecondaryClass = type === 'secondary' && props.default ? 'itsa-btn--default' : '';
	const className = ['itsa-btn', sizeClass, variantClass, defaultSecondaryClass, disabledClass].filter(Boolean).join(' ');
	const antdType: 'primary' | 'default' = type === 'primary' ? 'primary' : 'default';

	return (
		<ButtonAntd className={className} size={size} type={antdType} disabled={disabled} onClick={onClick}>
			{title}
		</ButtonAntd>
	);
};
