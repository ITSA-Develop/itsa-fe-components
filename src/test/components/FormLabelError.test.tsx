import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { FormLabelError } from '../../components/FormLabelError';

describe('FormLabelError component', () => {
    it('renders error label text', () => {
        const { container } = render(<FormLabelError label="Campo requerido" />);
        expect(screen.getByText('Campo requerido')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('sets htmlFor and id when provided', () => {
        render(<FormLabelError label="Error" htmlFor="field" id="field-error" />);
        const label = screen.getByText('Error');
        expect(label).toHaveAttribute('for', 'field');
        expect(label).toHaveAttribute('id', 'field-error');
        expect(label.className).toContain('text-red-500');
    });
});


