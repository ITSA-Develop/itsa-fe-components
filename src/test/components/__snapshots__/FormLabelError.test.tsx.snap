// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormLabelError component > renders error label text 1`] = `
<div>
  <label
    class="text-red-500"
  >
    Campo requerido
  </label>
</div>
`;
